File: medium/1861-rotating-the-box/solution.java
/**
 * 1861. Rotating the Box
 * Time: O(m × n)
 * Space: O(m × n)
 */
class Solution {
    public char[][] rotateTheBox(char[][] boxGrid) {
        int m = boxGrid.length;
        int n = boxGrid[0].length;

        for (int row = 0; row < m; row++) {
            int empty = n - 1;

            for (int col = n - 1; col >= 0; col--) {
                if (boxGrid[row][col] == '*') {
                    empty = col - 1;
                } else if (boxGrid[row][col] == '#') {
                    boxGrid[row][col] = '.';
                    boxGrid[row][empty] = '#';
                    empty--;
                }
            }
        }

        char[][] rotated = new char[n][m];

        for (int row = 0; row < m; row++) {
            for (int col = 0; col < n; col++) {
                rotated[col][m - 1 - row] = boxGrid[row][col];
            }
        }

        return rotated;
    }
}

File: medium/1861-rotating-the-box/solution.py
from typing import List


class Solution:
    def rotateTheBox(self, boxGrid: List[List[str]]) -> List[List[str]]:
        """
        1861. Rotating the Box
        Time: O(m × n)
        Space: O(m × n)
        """
        m, n = len(boxGrid), len(boxGrid[0])

        for row in range(m):
            empty = n - 1

            for col in range(n - 1, -1, -1):
                if boxGrid[row][col] == "*":
                    empty = col - 1
                elif boxGrid[row][col] == "#":
                    boxGrid[row][col] = "."
                    boxGrid[row][empty] = "#"
                    empty -= 1

        rotated = [["."] * m for _ in range(n)]

        for row in range(m):
            for col in range(n):
                rotated[col][m - 1 - row] = boxGrid[row][col]

        return rotated
